turbopuffer/namespace.py

```diff
@@ -30,6 +32,54 @@ def __init__(self, name: str, api_key: Optional[str] = None):
     def __str__(self) -> str:
         return f'tpuf-namespace:{self.name}'
 
+    def __eq__(self, other):
+        if isinstance(other, Namespace):
+            return self.name == other.name and self.backend == other.backend
+        else:
+            return False
+
+    def refresh_metadata(self):
+        response = self.backend.make_api_request('vectors', self.name, method='HEAD')
+        if response.get('status_code') == 200:
+            headers = response.get('headers', dict())
+            dimensions = int(headers.get('x-turbopuffer-dimensions', '0'))
+            approx_count = int(headers.get('x-turbopuffer-approx-num-vectors', '0'))
+            self.metadata = {
+                'exists': dimensions != 0,
+                'dimensions': dimensions,
+                'approx_count': approx_count,
+            }
+        else:
+            self.metadata = {
+                'exists': False,
+                'dimensions': 0,
+                'approx_count': 0,
+            }
+
+    def exists(self) -> bool:
+        """
+        Returns True if the namespace exists, and False if the namespace is missing or empty.
+        """
+        # Always refresh the exists check since metadata from list_namespaces() might be delayed.
+        self.refresh_metadata()
+        return self.metadata['exists']
+
+    def dimensions(self) -> int:
+        """
+        Returns the number of vector dimensions stored in this namespace.
+        """
+        if self.metadata is None:
+            self.refresh_metadata()
+        return self.metadata['dimensions']
+
+    def approx_count(self) -> int:
```

Comment:
When it's a method, it generally implies to me (if it's something that changes) that it'll do a request (i.e. not be cached). I'm not entirely sure what's better here, but I would suggest we always refresh metadata here and have an optional `cached=True` that allows stale entries. It would not be unusual to have the namespace in a global variable and call this every so often, but it would never change.

This doesn't matter for dimensions, but I think it matters for exists too.

In general, I like the design principle of "slow and not surprising" over "fast and surprising" 👍🏻

Reply:
The issue this tries to solve is that if you want to loop over all your namespaces and print size/dims that can take minutes with individual calls, vs instant when saving the metadata from the list endpoint.
Some customers definitely have more namespaces than the ~700 on the test account too, and this seems like the first thing everyone will do with this API.

I've attempted to invalidate what I can from the client, but an extra check to refresh on repeat calls would probably solve your concerns?

Reply:
I think the issue here is that the same type is used for iteration as well as one off construction.

When constructing a namespace and calling `approx_count` multiple times, I would assume that each call does a request.

When using the namespace list endpoint, I'd expect the count to be cached from the list call, and no further requests to be made.

It's definitely pragmatic to handle this by triggering a refresh on all calls, except the first call made on a Namespace constructed by the list call. It's a bit more magic than I'd hope for but also doesn't seem like it would cause anything really unexpected to happen.

Reply:
Using the same objects was very intentional so that you can query / list vectors / delete directly on the results rather than having to construct a new namespace object.

I'll implement the first-call caching like @pushrax described. I think the "magic" behavior of it is worth it to make the namespace listing more usable.